### PR TITLE
test: Start integrating pytest

### LINKED ISSUE #6503: Switch tests to py.test from nose
nose is unmaintained and recommends to use nose2, unittest or pytest

pytest has fixture based dependency injection, and parametrization, and is generally better

### PR DESCRIPTION
**Pull request by diru1100**
Reference #6503

#### Short description of what this resolves:
Converts test code to pytest standard

#### Changes proposed in this pull request:

Change in test code

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [ ] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

File: tests/all/unit/api/helpers/test_errors.py [hunk 1]
```diff
@@ -10,42 +8,29 @@
...
-        self.assertEqual(unprocessable_entity_error.status, 422)
-
-        # Bad Request Error
-        bad_request_error = BadRequestError({'source': ''}, 'Request cannot be served')
-        self.assertEqual(bad_request_error.status, 400)
+def test_error_response_dict_details():
+    """To test details in the form of dict"""
+    error_response = ErrorResponse(source="test source", detail="test detail")
+    expected_dict = {
+        'status': error_response.status,
```
[the remark this answers is not in the document]
Comment (diru1100) [reply tests/all/unit/api/helpers/test_errors.py]: Pre-commit ran black

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by diru1100** [comment]
@iamareebjamal  Can you please take a look at this?

**Comment by iamareebjamal** [comment]
This will close the issue when it just changes 1 test. Besides, the main tests we want to change are db tests

**Comment by kushthedude** [comment]
@diru1100 Are you certain that only one test file needs to be changed according to pytest style ? If not, Can you please remove the keyword `fixes` from the PR description .

**Comment by diru1100** [comment]
> This will close the issue when it just changes 1 test. Besides, the main tests we want to change are db tests

Yes, I just wanted to know if I was going in the right direction on this. I am planning on changing all the tests in a similar way in this PR itself.

**Comment by iamareebjamal** [comment]
Just change 1 DB test for now

File: tests/all/integration/api/helpers/test_auth.py [hunk 2]
```diff
@@ -42,12 +42,12 @@ def test_check_auth_admin(self):
...
             user.is_admin = True
             status = AuthManager.check_auth_admin('authtest@gmail.com', 'password')
-            self.assertEqual(True, status)
+            assert True == status
 
             user = create_user(email='authtest2@gmail.com', password='password')
             user.is_admin = False
             status = AuthManager.check_auth_admin('authtest2@gmail.com', 'password')
-            self.assertEqual(False, status)
+            assert False == status
```
[the remark this answers is not in the document]
Comment (diru1100) [reply tests/all/integration/api/helpers/test_auth.py]: doesn't apply here.

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

File: tests/all/integration/api/helpers/test_auth.py [hunk 3]
```diff
@@ -15,7 +15,7 @@ def test_load_user(self):
 
         with self.app.test_request_context():
             user = create_user(email='authtest@gmail.com', password='password')
-            self.assertEqual(user, db.session.query(User).get(user.id))
+            assert user == db.session.query(User).get(user.id)
```
Comment (iamareebjamal) [inline tests/all/integration/api/helpers/test_auth.py]: Nope. It is still not pytest test. This is standard unit test

File: tests/all/integration/api/helpers/test_auth.py [hunk 4]
```diff
@@ -1,54 +1,59 @@
-import unittest
-
+import pytest
+from flask import Flask
 from flask_login import login_user, logout_user
-
+from flask import Flask, session
```
Comment (niranjan94) [inline tests/all/integration/api/helpers/test_auth.py]: Codacy Issue found: [Reimport 'Flask' (imported line 2)](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

File: tests/all/integration/api/helpers/test_auth.py (hunk 4 above)
Comment (niranjan94) [inline tests/all/integration/api/helpers/test_auth.py]: Codacy Issue found: [Unused session imported from flask](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295)

File: tests/all/integration/api/helpers/test_auth.py [hunk 5]
```diff
@@ -1,54 +1,59 @@
-import unittest
-
+import pytest
+from flask import Flask
```
Comment (niranjan94) [inline tests/all/integration/api/helpers/test_auth.py]: Codacy Issue found: [Unused Flask imported from flask](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by diru1100** [comment]
@iamareebjamal @mrsaicharan1  can you please take a look at this?

**Comment by iamareebjamal** [comment]
Yes, I did. And the build is failing

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by diru1100** [comment]
image


The tests are passing locally when I run using pytest

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by diru1100** [comment]
@iamareebjamal  Can you please review this?
I think it's failing because travis build runs nosetest command.

**Comment by iamareebjamal** [comment]
Then change it to correct command

**Comment by diru1100** [comment]
do I have to add pytest in requirements and change travis yml file ?

**Comment by iamareebjamal** [comment]
You may

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

File: .travis.yml [hunk 6]
```diff
@@ -29,8 +29,9 @@ before_script:
   - export TEST_DATABASE_URL=postgresql://postgres@localhost:5432/test
   - bash scripts/test_multiple_heads.sh
   
-script:
-  - nosetests tests/ -v --with-coverage
+script: |
+  nosetests tests/ -v --with-coverage -I tests/all/integration/api/helpers/test_auth.py
+  pytest tests/all/integration/api/helpers/test_auth.py
```
Comment (iamareebjamal) [inline .travis.yml]: Not cool

File: .travis.yml [hunk 7]
```diff
@@ -40,3 +41,4 @@ branches:
   only:
     - master
     - development
+    - pytest
```
Comment (iamareebjamal) [inline .travis.yml]: Not cool

File: .travis.yml (hunk 6 above)
Comment (diru1100) [reply .travis.yml]: I didn't know another way to do it. 
Somehow I have to exclude test_auth from nostests and test it with pytest

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

File: .travis.yml [hunk 8]
```diff
@@ -29,8 +29,9 @@ before_script:
   - export TEST_DATABASE_URL=postgresql://postgres@localhost:5432/test
   - bash scripts/test_multiple_heads.sh
   
-script:
-  - nosetests tests/ -v --with-coverage
+script: 
+  - nosetests tests/ --ignore-files="test_auth\.py" -v --with-coverage
```
Comment (iamareebjamal) [inline .travis.yml]: Nope

Comment (diru1100) [reply .travis.yml]: sir, nosetests fails for a pytest code. Should I add a pytest script in yml file?

Comment (iamareebjamal) [reply .travis.yml]: All tests should run with 1 script

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/integration/api/validation/test_created_at.py  5
- tests/all/integration/api/helpers/test_auth.py  2
- tests/all/unit/api/helpers/test_errors.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/integration/api/helpers/test_auth.py  1
- tests/all/integration/api/validation/test_created_at.py  5
- tests/all/unit/api/helpers/test_errors.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  1
- tests/all/integration/api/validation/test_created_at.py  5
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/validation/test_created_at.py  5
- tests/all/integration/api/helpers/test_auth.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Comment by niranjan94** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- tests/all/integration/api/validation/test_created_at.py  5
- tests/all/unit/api/helpers/test_errors.py  5
- tests/all/integration/api/helpers/test_auth.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/fossasia/open-event-orga-server/pullRequest?prid=4970295&bid=16419772)

**Pull request merged**